directives/animate/animate.js

```diff
@@ -0,0 +1,251 @@
+					outerResolve();
+				}
+			}
+		};
+		this.clone.addEventListener('transitionend', onTransitionEnd);
+
+		Object.assign(this.clone.style, info.clone.start);
+		Object.assign(this.elem.style, info.elem.start);
+
+		if (info.elem.transition) {
+			await new Promise((r) => requestAnimationFrame(r));
+			Object.assign(this.elem.style, info.elem.transition);
+		}
+
+		await new Promise((r) => requestAnimationFrame(r));
+		Object.assign(this.clone.style, info.clone.end);
+		Object.assign(this.elem.style, info.elem.end);
+
+		return new Promise((resolve) => outerResolve = resolve);
+
+	}
+
+	async getElemInfo() {
+
+		await new Promise((r) => requestAnimationFrame(r));
+		const style = window.getComputedStyle(this.elem);
+
+		if (style.display !== 'flex') {
+			this.elem.style.display = 'grid'; // forces margins to collapse during size calculation
+		}
+
+		const rect = this.elem.getBoundingClientRect();
+
+		let top = 0;
+		let left = 0;
+		const offsetParent = this.elem.offsetParent;
+		let e = this.elem;
+		while (e && e !== document && !(e instanceof DocumentFragment)) {
+			if (e === offsetParent) {
+				break;
+			}
+			top += e.offsetTop - (e.scrollTop || 0);
+			left += e.offsetLeft - (e.scrollLeft || 0);
+			e = getComposedParent(e);
+		}
+
+		this.elem.style.removeProperty('display');
+
+		const paddingH = (parseInt(style.paddingLeft) || 0) + (parseInt(style.borderLeftWidth) || 0) + (parseInt(style.paddingRight) || 0) + (parseInt(style.borderRightWidth) || 0);
+		const paddingV = (parseInt(style.paddingTop) || 0) + (parseInt(style.borderTopWidth) || 0) + (parseInt(style.paddingBottom) || 0) + (parseInt(style.borderBottomWidth) || 0);
+		const marginsH = (parseInt(style.marginLeft) || 0) + (parseInt(style.marginRight) || 0);
+		const marginsV = (parseInt(style.marginTop) || 0) + (parseInt(style.marginBottom) || 0);
+
+		let cloneHeight = 0;
+		if (this.clone !== null) {
+			const cloneRect = this.clone.getBoundingClientRect();
+			cloneHeight = cloneRect.height;
+		}
+
+		return {
+			clone: {
+				height: cloneHeight,
+				fullHeight: rect.height + marginsV,
+				width: rect.width + marginsH
+			},
+			elem: {
+				height: rect.height - paddingV,
+				left,
+				opacity: this.state === 'showing' && this.clone === null ? '0' : style.opacity,
+				top,
+				width: rect.width - paddingH
+			}
+		};
+
+	}
+
+	async hide(opts) {
+
+		if (this.state === 'hiding' || this.state === 'hidden') {
+			return;
+		}
+		this.state = 'hiding';
+
+		// if focus is inside, move it to next focusable
+		const activeElem = getComposedActiveElement();
+		const focusIsInside = isComposedAncestor(this.elem, activeElem);
+		if (focusIsInside) {
+			const nextFocusable = getNextFocusable(activeElem);
+			if (nextFocusable) nextFocusable.focus();
+		}
+
+		if (reduceMotion || opts.skip === true) {
+			this.state = 'hidden';
+			this.elem.setAttribute('hidden', '');
+			return;
+		}
+
+		const info = await this.getElemInfo();
+
+		return this.animate({
+			clone: {
```

Comment:
Especially with a lot of elements, the content is overflowing during the hide transition. This was solved in the mixin by using `overflow: hidden` since the clone was wrapping the content, but in this case they are siblings... not sure what to do about that here. The `position: absolute` might also cause the content to have a brief period of overlap with other elements while the transition is running & content is overflowing.

Reply:
I've tried to capture this behaviour here, but had to slow the animation down 10x to make it really obvious:

![overflow](https://user-images.githubusercontent.com/5491151/110830072-d5318700-8266-11eb-9fe4-9f8cf86a8980.gif)

Andy, do you have an example of it looking weird at regular speed?

We could go back to making the clone a wrapper to solve this. My concern there would be then we'd definitely run into slotchange events with parents reacting to the expected children being removed and then re-added each time the animation runs. At least with the clone being a sibling, the actual element doesn't move in the DOM so they just need to ignore the clone.

Reply:
Yeah, you're right, I'm probably just nit-picking pretty hard, but I think it starts to get more noticeable here. The overall transition happens so quickly that it doesn't cause any issues and people wouldn't notice unless they really looked for it, though.
![animate directive 9](https://user-images.githubusercontent.com/77170518/110850921-58120c00-827e-11eb-877c-c6f8a57d6d18.gif)

If it's easier to solve the slotchange events issues by having them as siblings than with the clone as a wrapper, this overflow probably isn't big enough of an issue to warrant the change.

Reply:
Cool. Ok let's leave this for now as a known issue to watch for, and if it becomes a bigger problem we can weigh the options again.